### PR TITLE
feat(wasm-gen): add limits for recursions and heavy loops

### LINKED ISSUE #3313: Instrument loops in wasms generated with `wasm-gen` with breaks
Heavy loops are the main reason why `Not enough gas` error occurs during fuzz run. To avoid having this error frequently, an instrumentation for `loop` should be done.

### PR DESCRIPTION
**Pull request by StackOverflowExcept1on**
Resolves #3313

TODO:
- [x] tests on local machine
- [x] refactor code

### CODE REVIEW

**Comment by SabaunT** [comment]
please instrument loops too

**Comment by SabaunT** [comment]
Actually resolving #3313

File: utils/wasm-gen/src/config.rs [hunk 1]
```diff
@@ -165,7 +165,7 @@ impl<T> Default for StandardGearWasmConfigsBundle<T> {
         Self {
             log_info: Some("StandardGearWasmConfigsBundle".into()),
             existing_addresses: None,
-            remove_recursion: false,
+            remove_recursion: true,
```
Comment (SabaunT) [inline utils/wasm-gen/src/config.rs:168]: limit_recursions

Comment (SabaunT) [reply utils/wasm-gen/src/config.rs:168]: If this is set, then `gr_gas_available` must be defined in the config. The invariant must be implemented.

**Comment by breathx** [comment]
"[ ] test fuzzer on server"?

File: utils/wasm-gen/src/config.rs (hunk 1 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/config.rs]: I added `gr_gas_available` syscall to the config and renamed some things to `critical_gas_limit` in commit 4802c06ddacd8b0e1a0d896d62c95c060f42e0d9.

File: utils/wasm-gen/src/utils.rs [hunk 2]
```diff
@@ -215,6 +222,172 @@ fn find_recursion_impl<Callback>(
...
+                    new_instructions.push(Instruction::Call(call_index + 1));
+                }
+                _ => {
+                    new_instructions.push(instruction);
+                }
+            }
+        }
+    }
+
+    if rewrite_calls {
```
Comment (breathx) [inline utils/wasm-gen/src/utils.rs:354]: Arent we able to push call back?

**Review by breathx — APPROVED** [review]

File: utils/wasm-gen/src/utils.rs (hunk 2 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/utils.rs:354]: Not sure what you mean. We need to rewrite the sections after inserting the imports, just like the wasm-instrument does: https://github.com/gear-tech/wasm-instrument/blob/v0.2.1-sign-ext/src/gas_metering/mod.rs#L235

File: utils/wasm-gen/src/config.rs [hunk 3]
```diff
@@ -182,6 +195,12 @@ impl<T: Into<Hash>> ConfigsBundle for StandardGearWasmConfigsBundle<T> {
 
         let selectable_params = SelectableParams::default();
 
+        let mut injection_types = injection_types;
+        if critical_gas_limit.is_some() {
+            injection_types
+                .enable_syscall_import(InvocableSysCall::Loose(SysCallName::GasAvailable));
```
Comment (SabaunT) [inline utils/wasm-gen/src/config.rs:201]: I believe that must be done on a lower level when defining value in `GearWasmGeneratorConfigBuilder`.

File: utils/wasm-gen/src/tests.rs [hunk 4]
```diff
@@ -47,6 +47,40 @@ use std::{mem, num::NonZeroUsize};
 
 const UNSTRUCTURED_SIZE: usize = 1_000_000;
 
+#[test]
+fn inject_critical_gas_limit_works() {
```
Comment (SabaunT) [inline utils/wasm-gen/src/tests.rs:51]: please open an issue with P2 or P3 priority to introduce test that actually limiter's logic works as expected

File: utils/wasm-gen/src/utils.rs [hunk 5]
```diff
@@ -215,6 +222,172 @@ fn find_recursion_impl<Callback>(
     path.pop();
 }
 
+pub fn inject_critical_gas_limit(module: Module, critical_gas_limit: u64) -> Module {
```
Comment (SabaunT) [inline utils/wasm-gen/src/utils.rs:225]: the function is really huge. Can you slice it and add more docs?

**Review by SabaunT — COMMENTED** [review]
Awesome work done :rocket:

File: utils/wasm-gen/src/config.rs (hunk 3 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/config.rs]: done in d793944027c1b23af747be9f3ded0bb7f6562990

File: utils/wasm-gen/src/utils.rs (hunk 5 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/utils.rs:255]: done in 00bf3602a70ad63d0f7a1afa14297c52946765a9

File: utils/wasm-gen/src/tests.rs (hunk 4 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/tests.rs:51]: #3502

**Review by SabaunT — APPROVED** [review]

**Pull request merged**